### ISSUE TITLE
Discussion: Active Directory security group issues

### ISSUE THREAD

**Issue by slewis74** [OctopusDeploy/Issues#2879]
We've had a number of reports about issues with Users not being associated with Teams correctly in Octopus Deploy.  We've created this issue to discuss and better understand the various AD configurations where these issues appear.

The primary issue that's been reported seems to occur where the users and Octopus Deploy server belong to separate, trusted domains.

We have also recently had a report that this same issue occurs when users/groups are in a separate OU to the Octopus Deploy server.

**Comment by slewis74** [comment]
For anyone who wants to see how we're calling into AD, the code is now available in OctopusDeploy/DirectoryServicesAuthenticationProvider.

**Comment by slewis74** [comment]
The current implementation gets the groups a user is in as part of post login processing (and periodically when Session cookies are being refreshed, in case the user has been added to/removed from any groups) and stores them against the user.  Many of the issues seem to come back to the search for the user not succeeding and thus no groups being returned/assigned for the user.

This looks to be caused by a combination of the use of [PrincipalContext](https://msdn.microsoft.com/en-us/library/system.directoryservices.accountmanagement.principalcontext(v=vs.110).aspx) and [UserPrincipal.FindByName](https://msdn.microsoft.com/en-us/library/bb344891(v=vs.110).aspx).  A quick internet search will illustrate that this is commonly problematic when OU and trusted domains are involved.

An idea that's come to mind is to add an optional, alternate strategy where the server periodically requests the list of members from the groups that have been added to Teams in Octopus Deploy.  This is based on the assumption that the group assignment to Teams works as expected in the given environment.  We have would have the SIDs for the groups in this case so we could try using [GroupPrincipal.FindByIdentity](https://msdn.microsoft.com/en-us/library/bb344648.aspx), get the members and compare them to the list of known Users in Octopus Deploy to know who's in which Teams.

**Comment by thebigmoosey** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Discussion: Active Directory security group issues](https://trello.com/c/tFCXro9o/24-discussion-active-directory-security-group-issues)

**Comment by droyad** [comment]
@slewis74 Has this issue been resolved with the latest AD changes?

**Comment by soulechives** [comment]
@slewis74 Any updates on this issue? We've tested AD security group permissions without any luck.

**Comment by slewis74** [comment]
Apologies, I thought I'd comment on this and closed it. We did change a large amount of this behavior in 3.17, so include the background task to check the groups hourly instead of checking when the user logs in (we also load the groups when creating users of course).

In 3.17 we added the concept of Logins to a user, so you can specify the login details for a user separately to their username, display name etc. You can also have multiple logins now (e.g. AD and GoogleApps if you wanted to). Add an AD login lets you search in AD to ensure you're connecting to a known account.

If you're having trouble could you check whether the users have any AD logins listed? Could you also send some details about your AD setup and the task output from one of the **Synchronize External Security Group** tasks (that you should find are running each hour) to support@octopus.com and we'll try to get to the bottom of it.

**Comment by soulechives** [comment]
Hey Shannon,

Thanks for the reply! We would like to give 3.17 a shot. Is it fairly easy
to upgrade minor versions?

Can we run the installer on top of the existing installation?

Thanks!

Caleb

**Comment by slewis74** [comment]
Hi Caleb,
Yes the upgrade should be easy and you just run the MSI and it'll install over the existing binaries (your instances all remain in tact). There are some notes in our docs, https://octopus.com/docs/administration/upgrading

Key point on a major or minor upgrade, we may change things in the DB during these upgrades so it's worth having a recent backup just in case you want to roll back for any reason.

Shannon

**Comment by soulechives** [comment]
Hey Shannon,

Merry Christmas!

I was able to upgrade our Octopus server to version 3.17, but it did not
resolve the issue.

Our Octopus service is using a GMSA to run. The GMSA seems to have
appropriate permissions.

Through the Octopus API, I see that Octopus can see the AD group:

"Id": "Teams-82",
    "Name": "Test Team",
    "MemberUserIds": [],
    "ExternalSecurityGroups": [
      {
        "Id": "S-1-5-21-1343471177-3905618314-3112877782-6823",
        "DisplayName": "Test Group",
        "DisplayIdAndName": false
      }
    ],


There is an AD account which is a member of the Test Group.

There are permissions assigned to the Octopus team. I tested the
permissions by assigning an Octopus account to the team and they
worked.


Any ideas?


Thanks,


Caleb
